File: heron/lc212.java
public class Solution {

    private Set<String> resultSet = new HashSet<>();

    public List<String> findWords(char[][] board, String[] words) {

        // Builds the trie
        Trie trie = new Trie();
        for (String word : words) {
            trie.insert(word);
        }

        // Searches
        int m = board.length;
        int n = board[0].length;
        boolean[][] visited = new boolean[m][n];
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                searchAt(board, "", i, j, trie, m, n, visited);
            }
        }

        return new ArrayList<String>(resultSet);
    }

    private void searchAt(char[][] board, String word, int row, int col, Trie trie, int m, int n, boolean[][] visited) {

        // If out of bound or visted, stops here
        if (!((row >= 0 && row < m) && (col >= 0 && col < n)) || visited[row][col]) {
            return;
        }

        // Gets the new word
        word = word + board[row][col];

        // No word in dictionary starts with this string, stops here
        if (!trie.containsStartWith(word)) {
            return;
        }

        // If the dictionary contains the word we have so far, saves to the result set
        if (trie.contains(word)) {
            resultSet.add(word);
        }

        // Goes to next position from here
        visited[row][col] = true;
        searchAt(board, word, row + 1, col, trie, m, n, visited);
        searchAt(board, word, row - 1, col, trie, m, n, visited);
        searchAt(board, word, row, col + 1, trie, m, n, visited);
        searchAt(board, word, row, col - 1, trie, m, n, visited);
        visited[row][col] = false;

    }


    class TrieNode {
        public TrieNode[] children = new TrieNode[26];
        public String item = "";
    }

    class Trie {

        private TrieNode root = new TrieNode();

        public void insert(String word) {
            TrieNode node = root;
            for (char c : word.toCharArray()) {
                if (node.children[c - 'a'] == null) {
                    node.children[c - 'a'] = new TrieNode();
                }
                node = node.children[c - 'a'];
            }
            node.item = word;
        }

        public boolean containsStartWith(String prefix) {
            TrieNode node = root;
            for (char c : prefix.toCharArray()) {
                if (node.children[c - 'a'] == null) {
                    return false;
                }
                node = node.children[c - 'a'];
            }
            return true;
        }

        public boolean contains(String word) {
            TrieNode node = root;
            for (char c : word.toCharArray()) {
                if (node.children[c - 'a'] == null) {
                    return false;
                }
                node = node.children[c - 'a'];
            }
            return node.item.equals(word);
        }

    }

}

File: alice/LC212.py
class TrieNode:
    
    def __init__(self, character):
        self.character = character
        self.children = {}
        self.is_leaf = False

class Solution(object):
    def findWords(self, board, words):
        """
        :type board: List[List[str]]
        :type words: List[str]
        :rtype: List[str]
        """
        
        if not words or not board or not board[0]:
            return []

        root = TrieNode(None)

        for word in words:
            current_node = root
            for w in word:
                if w not in current_node.children:
                    temp_node = TrieNode(w)
                    current_node.children[w] = temp_node
                current_node = current_node.children[w]
            current_node.is_leaf = True

        self.out = set()
        self.N = len(board)
        self.M = len(board[0])
        self.board = board

        for i in xrange(self.N):
            for j in xrange(self.M):
                if self.board[i][j] in root.children:
                    # print i, j, self.board[i][j], root.children[self.board[i][j]].character
                    self.__dfs(i, j, self.board[i][j], root.children[self.board[i][j]], set())

        return list(self.out)

    def __dfs(self, i, j, prefix, node, seen):
        # print i, j, prefix, node, seen
        if node.is_leaf:
            # print prefix, self.board[i][j], node.character
            self.out.add(prefix)
        moves = [[0, 1], [0, -1], [1, 0], [-1, 0]]
        for move in moves:
            x = i + move[0]
            y = j + move[1]
            if 0 <= x < self.N and 0 <= y < self.M and (x, y) not in seen:
                next_char = self.board[x][y]
                
                # print i, j, x, y, next_char, node.children.keys()
                if next_char in node.children:
                    # if len(prefix) > 16: print x, y, prefix + next_char, node.children.keys()
                    seen.add((i, j))
                    self.__dfs(x, y, prefix + next_char, node.children[next_char], seen)
                    seen.remove((i, j))
